x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+                try {
+                    Base64.getUrlDecoder().decode(clientAuthorizationSharedSecret.toString());
+                } catch (Exception e) {
+                    throw new SettingsException(
+                        "Base64Url-encoding is required for the Client Authorization Shared Secret ["
+                            + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.CLIENT_AUTHORIZATION_SHARED_SECRET)
+                            + "]"
+                    );
+                }
+                break;
+            case JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_NONE:
+            default:
+                // If type is "None", the shared secret value must not be set
+                if (Strings.hasText(clientAuthorizationSharedSecret)) {
+                    throw new SettingsException(
+                        "Setting ["
+                            + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.CLIENT_AUTHORIZATION_SHARED_SECRET)
+                            + "] is not supported, because setting ["
+                            + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.CLIENT_AUTHORIZATION_TYPE)
+                            + "] is configured as ["
+                            + JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_NONE
+                            + "]"
+                    );
+                }
+                break;
+        }
+    }
+
+    // This logic is in a method for easy unit testing
+    public static void validateIssuerCredentialSettings(
+        final RealmConfig realmConfig,
+        final SecureString hmacSecretKey,
+        final String jwkSetPath,
+        final List<String> allowedSignatureAlgorithms
+    ) throws SettingsException {
+        final boolean hasHmacSecretKey = Strings.hasText(hmacSecretKey);
+        final boolean hasJwkSetPath = Strings.hasText(jwkSetPath);
+
+        // Validate HMAC or JWK Set are set, or both.
+        if ((hasHmacSecretKey == false) && (hasJwkSetPath == false)) {
+            throw new SettingsException(
+                "At least one setting is required for settings ["
+                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.ISSUER_HMAC_SECRET_KEY)
+                    + "] or ["
+                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
+                    + "]"
+            );
+        }
+
+        // Validate HMAC SecretKey Base64Url-encoding is OK, and decoded value is non-empty
+        if (hasHmacSecretKey) {
+            byte[] decodedHmacSecretKeyBytes = null;
+            try {
+                decodedHmacSecretKeyBytes = Base64.getUrlDecoder().decode(hmacSecretKey.toString());
+            } catch (Exception e) {
+                throw new SettingsException(
+                    "Base64Url decoding failed for setting ["
```

Comment:
I agree people get confused about Base64-regular, Base64-MIME, and Base64-URL. It would be easier if everything was just one format. I considered 3 options. I think you are advocating for 2. I ended up picking 3. My reasons for picking 3 are at the end.

**Option 1: Mixed Base64-URL, Base64-regular, and URL-safe**
- HTTP header JWT => Base64-URL
- HTTP header client secret => String
- ES keystore HMAC key => Base64-regular
- ES keystore client secret => String (exact match)

**Option 2: Mixed Base64-URL and Base64-regular**
- HTTP header JWT => Base64-URL
- HTTP header client secret => String
- ES keystore HMAC key => Base64-regular
- ES keystore client secret => Base64-regular (Important Assumption: Some URL-safe characters not safe for ES keystore)

**Option 3: Consistent Base64-URL for everything**
- HTTP header JWT => Base64-URL
- HTTP header client secret => Base64-URL
- ES keystore HMAC key => Base64-URL
- ES keystore client secret => Base64-URL (exact match)

Advantages of option 3 over 2:
1. Trivial setup: In option 3, one time generate and encode. The same value goes as-is into both HTTP headers and ES keystore. Customer can directly compare and diagnose mismatches themselves.
2. Support frequency: In option 2, imagine the customer picks a URL-safe string. They put it as-is into HTTP headers and ES keystore; they missed a step to Base64-regular encode before storing in ES keystore. When JWT realm decodes from ES keystore, or might be valid Base64-regular encoding. It will be decoded, compare fails, and we don't log the difference. Customer compares header and ES keystore values, sees they are identical, and opens a bug.

Option 3 seemed like the better option. It all hinges on one assumption, that some URL-safe characters are potentially unsafe for storage in ES keystore, hence they would need to be Base64-regular encoded for storing it. If ES keystore can handle every possible URL-safe character, then relaxing that constraint in option 2 would make it viable.

Reply:
TODO I will change to URL-safe String in HTTP requests and ES keystore.

Reply:
Done